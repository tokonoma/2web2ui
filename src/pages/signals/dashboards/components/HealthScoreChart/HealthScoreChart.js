/* eslint-disable lodash/prop-shorthand */
import React, { useState, Fragment } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import _ from 'lodash';
import { getCaretProps, getDoD } from 'src/helpers/signals';
import { selectCurrentHealthScoreDashboard } from 'src/selectors/signals';
import { Panel } from 'src/components/matchbox';
import BarChart from '../../../components/charts/barchart/BarChart';
import TooltipMetric from 'src/components/charts/TooltipMetric';
import { PanelLoading } from 'src/components';
import Callout from 'src/components/callout';
import useHibanaOverride from 'src/hooks/useHibanaOverride';
import MetricDisplay from '../MetricDisplay/MetricDisplay';
import { formatNumber, roundToPlaces } from 'src/helpers/units';
import thresholds from '../../../constants/healthScoreThresholds';
import {
  newModelLine,
  newModelMarginsHealthScore,
} from 'src/pages/signals/constants/healthScoreV2';
import { formatDate, getDateTicks } from 'src/helpers/date';
import OGStyles from './HealthScoreChart.module.scss';
import hibanaStyles from './HealthScoreChartHibana.module.scss';

export function HealthScoreChart(props) {
  const styles = useHibanaOverride(OGStyles, hibanaStyles);
  const [hoveredDate, setHovered] = useState(props.defaultHovered);

  function handleDateHover(props) {
    setHovered(props.date);
  }

  const { history = [], loading, error, filters } = props;
  const noData = history ? !_.some(getHealthScores()) : true;

  if (loading) {
    return <PanelLoading minHeight="407px" />;
  }

  if (error) {
    return (
      <Panel sectioned>
        <div className={styles.Content}>
          <div className={styles.ErrorMessage}>
            <Callout title="Unable to Load Data" height="auto" children={error.message} />
          </div>
        </div>
      </Panel>
    );
  }

  function getHealthScores() {
    return _.map(history, entry => entry.health_score);
  }

  function getXAxisProps() {
    const xTicks = getDateTicks(filters);
    return {
      ticks: xTicks,
      tickFormatter: tick => moment(tick).format('M/D'),
    };
  }

  function getTotalInjectionProps() {
    const injectionEntries = _.filter(history, entry => entry.injections);
    const injectionTotals = _.map(injectionEntries, entry => entry.injections);
    const sumInjectionTotals = _.sum(injectionTotals);
    return { value: _.isNil(sumInjectionTotals) ? 'n/a' : formatNumber(sumInjectionTotals) };
  }

  function getHoverInjectionProps() {
    const currentEntry = _.find(history, ['date', hoveredDate]);
    const value = currentEntry ? currentEntry.injections : null;
    return { value: _.isNil(value) ? 'n/a' : formatNumber(value) };
  }

  function getHoverDoDProps() {
    const previousDay = moment(hoveredDate)
      .subtract(1, 'day')
      .format('YYYY-MM-DD');
    const currentEntry = _.find(history, ['date', hoveredDate]);
    const prevEntry = _.find(history, ['date', previousDay]);
    const currentScore = currentEntry ? currentEntry.health_score : null;
    const prevScore = prevEntry ? prevEntry.health_score : null;
    const value = getDoD(currentScore, prevScore);
    return { value: _.isNil(value) ? 'n/a' : `${value}%`, ...getCaretProps(value) };
  }

  function getGap() {
    return history.length > 15 ? 0.2 : 1;
  }

  function getMin() {
    const min = _.min(getHealthScores());
    return { value: _.isNil(min) ? 'n/a' : min };
  }

  function getMax() {
    const max = _.max(getHealthScores());
    return { value: _.isNil(max) ? 'n/a' : max };
  }

  return (
    <Panel sectioned title={`${formatDate(filters.from)} â€“ ${formatDate(filters.to)}`}>
      <div className={styles.Content}>
        {noData && <Callout height="100%">Health Scores Not Available</Callout>}

        {!noData && (
          <Fragment>
            <BarChart
              margin={newModelMarginsHealthScore}
              gap={getGap()}
              onMouseOver={handleDateHover}
              onMouseOut={() => setHovered({})}
              hovered={hoveredDate}
              timeSeries={history}
              tooltipContent={({ payload = {} }) =>
                payload.ranking && (
                  <TooltipMetric
                    label="Health Score"
                    color={thresholds[payload.ranking].color}
                    value={`${roundToPlaces(payload.health_score, 1)}`}
                  />
                )
              }
              yAxisRefLines={[
                { y: 80, stroke: thresholds.good.color, strokeWidth: 1 },
                { y: 55, stroke: thresholds.danger.color, strokeWidth: 1 },
              ]}
              xAxisRefLines={newModelLine}
              yKey="health_score"
              xAxisProps={getXAxisProps()}
              yAxisProps={{ ticks: [0, 55, 80, 100] }}
            />
          </Fragment>
        )}
        <div className={styles.Metrics}>
          <MetricDisplay
            data-id="health-score-total-injections"
            label="Total Injections"
            {...getTotalInjectionProps()}
          />
          <MetricDisplay data-id="health-score-high-value" label="High" {...getMax()} />
          <MetricDisplay data-id="health-score-low-value" label="Low" {...getMin()} />
          {typeof hoveredDate === 'string' && (
            <Fragment>
              <div className={styles.Divider} />
              <div className={styles.Injections}>
                <MetricDisplay label="Injections" {...getHoverInjectionProps()} />
              </div>
              <div className={styles.DoDChange}>
                <MetricDisplay label="DoD Change" {...getHoverDoDProps()} />
              </div>
            </Fragment>
          )}
        </div>
      </div>
    </Panel>
  );
}

const mapStateToProps = state => ({
  filters: state.signalOptions,
  ...selectCurrentHealthScoreDashboard(state),
});

export default connect(mapStateToProps, {})(HealthScoreChart);
