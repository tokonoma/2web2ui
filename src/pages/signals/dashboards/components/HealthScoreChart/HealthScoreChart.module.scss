@import '~@sparkpost/matchbox/src/styles/config.scss';

.Content {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 100%;
  height: 375px;
}

.ErrorMessage {
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;
}

.Metrics {
  text-align: center;
  display: flex;
  justify-content: left;
}

.Injections,
.DoDChange {
  width: rem(140);
  text-align: left;
}

.Divider {
  margin-top: rem(10);
  height: 60%;
  margin-right: spacing();
  width: 1px;
  background: color(gray, 6);
}

.xDividerLabel {
  fill: color(gray, base);
  font-size: rem(12);
  line-height: line-height(500);
  text-anchor: start;
}
