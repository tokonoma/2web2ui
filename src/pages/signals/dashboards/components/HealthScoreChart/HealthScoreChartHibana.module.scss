@import '~@sparkpost/design-tokens-hibana/tokens';

.Injections,
.DodDChange {
  width: 140px;
}

.Divider {
  margin-top: spacing(300);
  margin-right: spacing(500);
  background: color(gray, 300);
}

.xDividerLabel {
  fill: color(gray, 700);
  font-size: font-size(100);
  line-height: line-height(500);
}
