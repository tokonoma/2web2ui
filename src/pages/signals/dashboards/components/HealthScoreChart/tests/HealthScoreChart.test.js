import React from 'react';
import { shallow } from 'enzyme';
import { HealthScoreChart } from '../HealthScoreChart';

jest.mock('src/hooks/useHibanaOverride', () => styles => styles);

describe('Signals Health Score Chart', () => {
  let props;
  let subject;

  beforeEach(() => {
    props = {
      loading: false,
      error: null,
      current_health_score: 88,
      WoW: -5,
      current_DoD: 5,
      history: [
        {
          date: '2019-03-24',
          health_score: 75,
          injections: 75000000000,
          ranking: 'warning',
        },
        {
          date: '2019-03-25',
          health_score: 96,
          injections: 75000000000,
          ranking: 'good',
        },
        {
          date: '2019-03-26',
          health_score: 23,
          injections: 75000000000,
          ranking: 'danger',
        },
        {
          date: '2019-03-27',
          health_score: null,
          ranking: null,
        },
      ],
      filters: {
        from: '2014-12-01',
        to: '2015-02-01',
      },
    };

    subject = (options = {}) => shallow(<HealthScoreChart {...props} {...options} />);
  });

  it('renders happy path correctly', () => {
    expect(subject(props)).toMatchSnapshot();
  });

  it('renders loading correctly', () => {
    const wrapper = subject({ loading: true });
    expect(wrapper.find('PanelLoading')).toExist();
  });

  it('renders error correctly', () => {
    const wrapper = subject({ error: { message: 'mock error' } });
    expect(wrapper.find('Callout')).toMatchSnapshot();
  });

  it('renders no account level data callout correctly', () => {
    props.history = [];
    const wrapper = subject(props);
    expect(wrapper.find('Callout')).toMatchSnapshot();
  });

  it('renders metric displays', () => {
    const wrapper = subject({ defaultHovered: '2019-03-26' });
    expect(wrapper.find('MetricDisplay').at(0)).toMatchSnapshot();
    expect(wrapper.find('MetricDisplay').at(1)).toMatchSnapshot();
  });
});
