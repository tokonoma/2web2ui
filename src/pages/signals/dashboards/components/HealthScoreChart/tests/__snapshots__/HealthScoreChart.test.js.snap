// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signals Health Score Chart renders error correctly 1`] = `
<Callout
  height="auto"
  title="Unable to Load Data"
>
  mock error
</Callout>
`;

exports[`Signals Health Score Chart renders happy path correctly 1`] = `
<Panel
  sectioned={true}
  title="Dec 1 2014 â€“ Feb 1 2015"
>
  <div
    className="Content"
  >
    <BarChart
      gap={1}
      height={250}
      margin={
        Object {
          "bottom": 5,
          "left": 18,
          "right": 8,
          "top": 21,
        }
      }
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      shouldHighlightSelected={true}
      timeSeries={
        Array [
          Object {
            "date": "2019-03-24",
            "health_score": 75,
            "injections": 75000000000,
            "ranking": "warning",
          },
          Object {
            "date": "2019-03-25",
            "health_score": 96,
            "injections": 75000000000,
            "ranking": "good",
          },
          Object {
            "date": "2019-03-26",
            "health_score": 23,
            "injections": 75000000000,
            "ranking": "danger",
          },
          Object {
            "date": "2019-03-27",
            "health_score": null,
            "ranking": null,
          },
        ]
      }
      tooltipContent={[Function]}
      width="99%"
      xAxisProps={
        Object {
          "tickFormatter": [Function],
          "ticks": Array [
            "2014-12-01",
            "2015-01-01",
            "2015-02-01",
          ],
        }
      }
      xAxisRefLines={
        Array [
          Object {
            "label": Object {
              "className": "xDividerLabel",
              "offset": 14,
              "position": "top",
              "value": "New",
            },
            "stroke": "#d2d2d7",
            "strokeWidth": 1,
            "x": "2019-06-07",
          },
          Object {
            "label": Object {
              "className": "xDividerLabel",
              "position": "top",
              "value": "Model",
            },
            "strokeWidth": 0,
            "x": "2019-06-07",
          },
        ]
      }
      xKey="date"
      yAxisProps={
        Object {
          "ticks": Array [
            0,
            55,
            80,
            100,
          ],
        }
      }
      yAxisRefLines={
        Array [
          Object {
            "stroke": "#067852",
            "strokeWidth": 1,
            "y": 80,
          },
          Object {
            "stroke": "#d9363e",
            "strokeWidth": 1,
            "y": 55,
          },
        ]
      }
      yKey="health_score"
      yRange={
        Array [
          "auto",
          "auto",
        ]
      }
    />
    <div
      className="Metrics"
    >
      <MetricDisplay
        data-id="health-score-total-injections"
        label="Total Injections"
        value="225B"
      />
      <MetricDisplay
        data-id="health-score-high-value"
        label="High"
        value={96}
      />
      <MetricDisplay
        data-id="health-score-low-value"
        label="Low"
        value={23}
      />
    </div>
  </div>
</Panel>
`;

exports[`Signals Health Score Chart renders metric displays 1`] = `
<MetricDisplay
  data-id="health-score-total-injections"
  label="Total Injections"
  value="225B"
/>
`;

exports[`Signals Health Score Chart renders metric displays 2`] = `
<MetricDisplay
  data-id="health-score-high-value"
  label="High"
  value={96}
/>
`;

exports[`Signals Health Score Chart renders no account level data callout correctly 1`] = `
<Callout
  height="100%"
>
  Health Scores Not Available
</Callout>
`;
